File: MusicFestivalMPP/repository/RepositoryUser.cs
using System;
using System.Collections.Generic;
using System.Data;
using System.Data.SQLite;
using log4net;
using log4net.Util;
using MusicFestivalMPP.model;
using MusicFestivalMPP.repository.interfaces;

namespace MusicFestivalMPP.repository;

public class RepositoryUser: IRepositoryUser
{

    private static readonly ILog Log = LogManager.GetLogger("RepositoryUser");

    private IDictionary<String, string> properties;

    public RepositoryUser(IDictionary<string, string> properties)
    {
        Log.Info("Creating RepositoryUser.");
        this.properties = properties;
    }

    public User Save(User entity)
    {
        Log.InfoFormat("Saving user: {0}.", entity);
        IDbConnection connection = DbUtils.GetConnection(properties);
        try
        {
            using (IDbCommand command = connection.CreateCommand())
            {
                command.CommandText = "insert into users (username, password) values (@username, @password)";
                IDbDataParameter parameterUsername = command.CreateParameter();
                parameterUsername.ParameterName = "@username";
                parameterUsername.Value = entity.Username;
                command.Parameters.Add(parameterUsername);
                IDbDataParameter parameterPassword = command.CreateParameter();
                parameterPassword.ParameterName = "@password";
                parameterPassword.Value = entity.Password;
                command.Parameters.Add(parameterPassword);
                int result = command.ExecuteNonQuery();
                Log.InfoFormat("Saved {0} instances.", result);
                return entity;
            }
        }
        catch (SQLiteException e)
        {
            Log.Error(e);
        }
        return null;
    }

    public User Find(long id)
    {
        Log.InfoFormat("Finding user with ID: {0}.", id);
        IDbConnection connection = DbUtils.GetConnection(properties);
        try
        {
            using (IDbCommand command = connection.CreateCommand())
            {
                command.CommandText = "select username, password from users where user_id=@id";
                IDbDataParameter paramId = command.CreateParameter();
                paramId.ParameterName = "@id";
                paramId.Value = id;
                command.Parameters.Add(paramId);
                using (var result = command.ExecuteReader())
                {
                    if (result.Read())
                    {
                        String username = result.GetString(0);
                        String password = result.GetString(1);
                        User user = new User(id, username, password);
                        Log.InfoFormat("Found user: {0}.", user);
                        return user;
                    }
                }
            }
        }
        catch (SQLiteException e)
        {
            Log.Error(e);
        }
        Log.InfoFormat("Didn't find the user with ID: {0}", id);
        return null;
    }

    public User Delete(User entity)
    {
        Log.InfoFormat("Deleting user: {0}.", entity);
        IDbConnection connection = DbUtils.GetConnection(properties);
        try
        {
            using (IDbCommand command = connection.CreateCommand())
            {
                command.CommandText = "delete from users where user_id = @id";
                IDbDataParameter parameterId = command.CreateParameter();
                parameterId.ParameterName = "@id";
                parameterId.Value = entity.Id;
                command.Parameters.Add(parameterId);
                int result = command.ExecuteNonQuery();
                Log.InfoFormat("Deleted {0} instances.", result);
                return entity;
            }
        }
        catch (SQLiteException e)
        {
            Log.Error(e);
        }
        return null;
    }

    public User Update(User entity)
    {
        Log.InfoFormat("Updating user: {0}.", entity);
        IDbConnection connection = DbUtils.GetConnection(properties);
        try
        {
            using (IDbCommand command = connection.CreateCommand())
            {
                command.CommandText = "update users set username = @username, password = @password where user_id = @id";
                IDbDataParameter parameterUsername = command.CreateParameter();
                parameterUsername.ParameterName = "@username";
                parameterUsername.Value = entity.Username;
                command.Parameters.Add(parameterUsername);
                IDbDataParameter parameterPassword = command.CreateParameter();
                parameterPassword.ParameterName = "@password";
                parameterPassword.Value = entity.Password;
                command.Parameters.Add(parameterPassword);
                IDbDataParameter parameterId= command.CreateParameter();
                parameterId.ParameterName = "@id";
                parameterId.Value = entity.Id;
                command.Parameters.Add(parameterId);
                int result = command.ExecuteNonQuery();
                Log.InfoFormat("Updated {0} instances.", result);
                return entity;
            }
        }
        catch (SQLiteException e)
        {
            Log.Error(e);
        }
        return null;
    }

    public IList<User> FindAll()
    {
        Log.Info("Finding all users");
        IDbConnection connection = DbUtils.GetConnection(properties);
        IList<User> users = new List<User>();
        try
        {
            using (var command = connection.CreateCommand())
            {
                command.CommandText = "select user_id, username, password from users";
                using (var result = command.ExecuteReader())
                {
                    while (result.Read())
                    {
                        int userId = result.GetInt32(0);
                        String username = result.GetString(1);
                        String password = result.GetString(2);
                        User user = new User(userId, username, password);
                        users.Add(user);
                    }
                }
            }
        }
        catch (SQLiteException e)
        {
            Log.Error(e);
        }
        Log.InfoExt("Exit FindAll.");
        return users;
    }

    public User FindUserByUsername(string username)
    {
        Log.InfoFormat("Finding user with username: {0}.", username);
        IDbConnection connection = DbUtils.GetConnection(properties);
        try
        {
            using (IDbCommand command = connection.CreateCommand())
            {
                command.CommandText = "select user_id, password from users where username=@username";
                IDbDataParameter parameterUsername = command.CreateParameter();
                parameterUsername.ParameterName = "@username";
                parameterUsername.Value = username;
                command.Parameters.Add(parameterUsername);
                using (var result = command.ExecuteReader())
                {
                    if (result.Read())
                    {
                        int userId = result.GetInt32(0);
                        String password = result.GetString(1);
                        User user = new User(userId, username, password);
                        Log.InfoFormat("Found user: {0}.", user);
                        return user;
                    }
                }
            }
        }
        catch (SQLiteException e)
        {
            Log.Error(e);
        }
        Log.InfoFormat("Didn't find the user with username: {0}", username);
        return null;
    }

    public User CheckConnection(string username, string password)
    {
        Log.InfoFormat("Check connection for user with username: {0}.", username);
        IDbConnection connection = DbUtils.GetConnection(properties);
        try
        {
            using (IDbCommand command = connection.CreateCommand())
            {
                command.CommandText = "select user_id from users where username=@username and password = @password";
                IDbDataParameter parameterUsername = command.CreateParameter();
                parameterUsername.ParameterName = "@username";
                parameterUsername.Value = username;
                command.Parameters.Add(parameterUsername);
                IDbDataParameter parameterPassword = command.CreateParameter();
                parameterPassword.ParameterName = "@password";
                parameterPassword.Value = password;
                command.Parameters.Add(parameterPassword);
                using (var result = command.ExecuteReader())
                {
                    if (result.Read())
                    {
                        int userId = result.GetInt32(0);
                        User user = new User(userId, username, password);
                        Log.InfoFormat("Found user: {0}.", user);
                        return user;
                    }
                }
            }
        }
        catch (SQLiteException e)
        {
            Log.Error(e);
        }
        Log.InfoFormat("Didn't connect user with username: {0}", username);
        return null;
    }
}

File: code/src/main/java/repository/RepositoryUser.java
package repository;

import model.User;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import repository.interfaces.RepositoryInterfaceUser;
import util.JDBCUtils;

import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.util.ArrayList;
import java.util.List;
import java.util.Optional;
import java.util.Properties;

public class RepositoryUser implements RepositoryInterfaceUser {

    private JDBCUtils jdbcUtils;

    private static final Logger logger = LogManager.getLogger();

    public RepositoryUser(Properties properties) {
        logger.info("Initializing RepositoryUser with properties: {} ", properties);
        jdbcUtils = new JDBCUtils(properties);
    }

    @Override
    public Optional<User> save(User user) {
        logger.traceEntry("Saving user: {} ", user);
        Connection connection = jdbcUtils.getConnection();
        try(PreparedStatement preparedStatement = connection.prepareStatement("insert into users (username, password) values (?, ?)")) {
            preparedStatement.setString(1, user.getUsername());
            preparedStatement.setString(2, user.getPassword());
            int result = preparedStatement.executeUpdate();
            logger.trace("Saved {} instances.", result);
        } catch (SQLException e) {
            logger.error(e);
            System.err.println("DB error: " + e);
            return Optional.empty();
        }
        logger.traceExit();
        return Optional.of(user);
    }

    @Override
    public Optional<User> find(Long ID) {
        logger.traceEntry("Finding user with ID: {}.", ID);
        Connection connection = jdbcUtils.getConnection();
        try(PreparedStatement preparedStatement = connection.prepareStatement("select * from users where user_id = ?")) {
            preparedStatement.setLong(1, ID);
            try(ResultSet resultSet = preparedStatement.executeQuery()) {
                if(resultSet.next()) {
                    String username = resultSet.getString("username");
                    String password = resultSet.getString("password");
                    User user = new User(ID, username, password);
                    logger.traceExit(user);
                    return Optional.of(user);
                }
            }
        } catch (SQLException e) {
            logger.error(e);
            System.err.println("DB error: " + e);
        }
        logger.traceExit();
        return Optional.empty();
    }

    @Override
    public Optional<User> delete(User user) {
        logger.traceEntry("Deleting user: {} ", user);
        Connection connection = jdbcUtils.getConnection();
        try(PreparedStatement preparedStatement = connection.prepareStatement("delete from users where user_id = ?")) {
            preparedStatement.setLong(1, user.getID());
            int result = preparedStatement.executeUpdate();
            logger.trace("Deleted {} instances.", result);
        } catch (SQLException e) {
            logger.error(e);
            System.err.println("DB error: " + e);
            return Optional.empty();
        }
        logger.traceExit();
        return Optional.of(user);
    }

    @Override
    public Optional<User> update(User user) {
        logger.traceEntry("Updating user: {} ", user);
        Connection connection = jdbcUtils.getConnection();
        try(PreparedStatement preparedStatement = connection.prepareStatement("update users set username = ?,  password = ? where user_id = ?")) {
            preparedStatement.setString(1, user.getUsername());
            preparedStatement.setString(2, user.getPassword());
            preparedStatement.setLong(3, user.getID());
            int result = preparedStatement.executeUpdate();
            logger.trace("Updated {} instances.", result);
        } catch (SQLException e) {
            logger.error(e);
            System.err.println("DB error: " + e);
            return Optional.empty();
        }
        logger.traceExit();
        return Optional.of(user);
    }

    @Override
    public List<User> findAll() {
        logger.traceEntry("Finding all users.");
        Connection connection = jdbcUtils.getConnection();
        List<User> users = new ArrayList<>();
        try(PreparedStatement preparedStatement = connection.prepareStatement("select * from users")) {
            try(ResultSet resultSet = preparedStatement.executeQuery()) {
                while(resultSet.next()) {
                    Long user_id = resultSet.getLong("user_id");
                    String username = resultSet.getString("username");
                    String password = resultSet.getString("password");
                    User user = new User(user_id, username, password);
                    users.add(user);
                }
            }
        } catch (SQLException e) {
            logger.error(e);
            System.err.println("DB error: " + e);
        }
        logger.traceExit();
        return users;
    }

    @Override
    public Optional<User> findUserByUsername(String username) {
        logger.traceEntry("Finding user with username: {}", username);
        Connection connection = jdbcUtils.getConnection();
        try(PreparedStatement preparedStatement = connection.prepareStatement("select * from users where username = ?")) {
            preparedStatement.setString(1, username);
            try(ResultSet resultSet = preparedStatement.executeQuery()) {
                if(resultSet.next()) {
                    Long user_id = resultSet.getLong("user_id");
                    String password = resultSet.getString("password");
                    User user = new User(user_id, username, password);
                    logger.traceExit(user);
                    return Optional.of(user);
                }
            }
        } catch (SQLException e) {
            logger.error(e);
            System.err.println("DB error: " + e);
        }
        logger.traceExit();
        return Optional.empty();
    }

    @Override
    public Optional<User> checkConnection(String username, String password) {
        logger.traceEntry("Check connection for user with username: {}", username);
        Connection connection = jdbcUtils.getConnection();
        try(PreparedStatement preparedStatement = connection.prepareStatement("select * from users where username = ? and password = ?")) {
            preparedStatement.setString(1, username);
            preparedStatement.setString(2, password);
            try(ResultSet resultSet = preparedStatement.executeQuery()) {
                if(resultSet.next()) {
                    Long user_id = resultSet.getLong("user_id");
                    User user = new User(user_id, username, password);
                    logger.traceExit(user);
                    return Optional.of(user);
                }
            }
        } catch (SQLException e) {
            logger.error(e);
            System.err.println("DB error: " + e);
        }
        logger.traceExit();
        return Optional.empty();
    }
}
